File: core/src/nl/t64/game/rpg/PreferenceManager.java
package nl.t64.game.rpg;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Graphics;
import com.badlogic.gdx.Preferences;
import lombok.Getter;
import nl.t64.game.rpg.constants.Constant;

import java.util.Arrays;
import java.util.Comparator;
import java.util.NoSuchElementException;


public class PreferenceManager {

    private static final String SETTINGS_FILE = "settings.dat";
    private static final String SETTING_SCREEN = "isFullscreen";
    private static final String SETTING_MUSIC = "isMusicOn";
    private static final String SETTING_SOUND = "isSoundOn";
    private static final String SETTING_DEBUG = "isInDebugMode";
    private static final boolean FULLSCREEN_DEFAULT = true;
    private static final boolean MUSIC_DEFAULT = true;
    private static final boolean SOUND_DEFAULT = true;
    private static final boolean DEBUG_MODE_DEFAULT = false;

    private final Preferences preferences;
    @Getter
    private boolean isFullscreen;
    @Getter
    private boolean isMusicOn;
    @Getter
    private boolean isSoundOn;
    @Getter
    private boolean isInDebugMode;

    PreferenceManager() {
        this.preferences = Gdx.app.getPreferences(SETTINGS_FILE);
        this.isFullscreen = this.preferences.getBoolean(SETTING_SCREEN, FULLSCREEN_DEFAULT);
        this.isMusicOn = this.preferences.getBoolean(SETTING_MUSIC, MUSIC_DEFAULT);
        this.isSoundOn = this.preferences.getBoolean(SETTING_SOUND, SOUND_DEFAULT);
        this.isInDebugMode = DEBUG_MODE_DEFAULT;
    }

    public void toggleFullscreen() {
        if (Gdx.graphics.isFullscreen()) {
            setWindowedMode();
        } else {
            setFullscreenMode();
        }
        isFullscreen = Gdx.graphics.isFullscreen();
        preferences.putBoolean(SETTING_SCREEN, isFullscreen);
        preferences.flush();
    }

    public void toggleMusic(boolean mustPlayBgmImmediately) {
        isMusicOn ^= true;
        Utils.getAudioManager().toggleMusic(isMusicOn, mustPlayBgmImmediately);
        preferences.putBoolean(SETTING_MUSIC, isMusicOn);
        preferences.flush();
    }

    public void toggleSound() {
        isSoundOn ^= true;
        Utils.getAudioManager().toggleSound(isSoundOn);
        preferences.putBoolean(SETTING_SOUND, isSoundOn);
        preferences.flush();
    }

    public void toggleDebugMode() {
        isInDebugMode ^= true;
    }

    private void setWindowedMode() {
        Gdx.graphics.setWindowedMode(Constant.SCREEN_WIDTH, Constant.SCREEN_HEIGHT);
    }

    private void setFullscreenMode() {
        Graphics.DisplayMode displayMode = Arrays.stream(Gdx.graphics.getDisplayModes())
                                                 .filter(this::hasGivenResolution)
                                                 .max(Comparator.comparing(this::getRefreshRate))
                                                 .orElseThrow(NoSuchElementException::new);
        Gdx.graphics.setFullscreenMode(displayMode);
    }

    private boolean hasGivenResolution(Graphics.DisplayMode mode) {
        return mode.width == Constant.SCREEN_WIDTH
               && mode.height == Constant.SCREEN_HEIGHT;
    }

    private int getRefreshRate(Graphics.DisplayMode mode) {
        return mode.refreshRate;
    }

}

File: core/src/nl/t64/game/rpg/PreferenceManager.kt
package nl.t64.game.rpg

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Graphics
import com.badlogic.gdx.Preferences
import ktx.preferences.get
import ktx.preferences.set
import nl.t64.game.rpg.Utils.audioManager
import nl.t64.game.rpg.constants.Constant

private const val SETTINGS_FILE = "settings.dat"
private const val SETTING_SCREEN = "isFullscreen"
private const val SETTING_MUSIC = "isMusicOn"
private const val SETTING_SOUND = "isSoundOn"
private const val SETTING_DEBUG = "isInDebugMode"
private const val FULLSCREEN_DEFAULT = true
private const val MUSIC_DEFAULT = true
private const val SOUND_DEFAULT = true
private const val DEBUG_MODE_DEFAULT = false

class PreferenceManager {

    private val preferences: Preferences = Gdx.app.getPreferences(SETTINGS_FILE)
    var isFullscreen = preferences[SETTING_SCREEN, FULLSCREEN_DEFAULT]
    var isMusicOn = preferences[SETTING_MUSIC, MUSIC_DEFAULT]
    var isSoundOn = preferences[SETTING_SOUND, SOUND_DEFAULT]
    var isInDebugMode = DEBUG_MODE_DEFAULT

    fun toggleFullscreen() {
        if (Gdx.graphics.isFullscreen) {
            setWindowedMode()
        } else {
            setFullscreenMode()
        }
        isFullscreen = Gdx.graphics.isFullscreen
        preferences[SETTING_SCREEN] = isFullscreen
        preferences.flush()
    }

    fun setFullscreenAccordingToPreference() {
        if (isFullscreen) setFullscreenMode()
    }

    fun toggleMusic(mustPlayBgmImmediately: Boolean) {
        isMusicOn = isMusicOn.not()
        audioManager.toggleMusic(isMusicOn, mustPlayBgmImmediately)
        preferences[SETTING_MUSIC] = isMusicOn
        preferences.flush()
    }

    fun toggleSound() {
        isSoundOn = isSoundOn.not()
        audioManager.toggleSound(isSoundOn)
        preferences[SETTING_SOUND] = isSoundOn
        preferences.flush()
    }

    fun toggleDebugMode() {
        isInDebugMode = isInDebugMode.not()
    }

    private fun setWindowedMode() {
        Gdx.graphics.setWindowedMode(Constant.SCREEN_WIDTH, Constant.SCREEN_HEIGHT)
    }

    private fun setFullscreenMode() {
        val displayMode = Gdx.graphics.displayModes
            .filter { hasGivenResolution(it) }
            .maxByOrNull { getRefreshRate(it) }
            ?: throw NoSuchElementException()
        Gdx.graphics.setFullscreenMode(displayMode)
    }

    private fun hasGivenResolution(mode: Graphics.DisplayMode): Boolean {
        return (mode.width == Constant.SCREEN_WIDTH
                && mode.height == Constant.SCREEN_HEIGHT)
    }

    private fun getRefreshRate(mode: Graphics.DisplayMode): Int {
        return mode.refreshRate
    }

}
